k8s/13.md

```diff
@@ -0,0 +1,125 @@
+# StatefulSet
+
+## Table of contents
+
+- [Changes to Helm chart](#changes-to-helm)
+- [Testing the service](#testing-the-service)
+- Bonus part
+  - [Second app chart updates](#second-app-chart-updates)
+  - [Update strategies](#update-strategies)
+
+## Changes to Helm
+
+- Deployment is replaced with StatefulSet
+- `replicaCount` changed from 1 to 2
+- `helm install --dry-run --debug` reported no errors
+- nothing was moved to `values.yaml` since there is nothing to move
+- no **PersistentVolumeClaims** added since `redis` is used for persistency.
+
+`kubectl get po,sts,svc,pvc` output:
+
+```text
+NAME                                READY   STATUS    RESTARTS   AGE
+pod/py-app-python-app-0             1/1     Running   0          49s
+pod/py-app-python-app-1             1/1     Running   0          28s
+pod/py-app-redis-7568c9bf5c-nzd57   1/1     Running   0          7m43s
+
+NAME                                 READY   AGE
+statefulset.apps/py-app-python-app   2/2     49s
+
+NAME                        TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
+service/kubernetes          ClusterIP   10.96.0.1      <none>        443/TCP    28d
+service/py-app-python-app   ClusterIP   10.102.216.8   <none>        8000/TCP   7m43s
+service/py-app-redis        ClusterIP   10.96.119.45   <none>        6379/TCP   7m43s
+```
+
+## Testing the service
+
+After running `minikube service py-app-python-app`, the service was accessed several times. However, the number of visits reported at `/status` did not change between two replicas — this is because they both use a single `redis` instance.
+
+![image](https://user-images.githubusercontent.com/29694249/204130593-1b926fae-39e7-4079-b5df-72ac7a1a80df.png)
+
+To see that both pods are actually in use, I configured `gunicorn` to print access log entries to stdout. This way we will be able to see these through `kubectl logs` command.
+
+Using some bash tricks we can find the number of requests served by each pod:
+
+```sh
+for i in 0 1
+do
+  POD="pod/py-app-python-app-$i"
+  echo -n "$POD: "
+  kubectl logs "$POD" | grep "GET / HTTP" | wc -l
+done
+```
+
+Output:
+
+```text
+pod/py-app-python-app-0: 42
+pod/py-app-python-app-1: 33
+```
+
+As we see, services are not distributing the requests equally. Plus, StatefulSet launched the pods sequentially, so the first pod was able to handle requests earlier than the second one.
+
+## Launch in parallel
+
+Since the pods don't depend on each other, we can actually configure StatefulSet to launch the pods in parallel by adding `podManagementPolicy: "Parallel"` to the spec. I decided to add it to `values.yaml` and reference in the StatefulSet template:
+
+```diff
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ include "python-app.fullname" . }}
+  labels:
+    {{- include "python-app.labels" . | nindent 4 }}
+spec:
++ podManagementPolicy: {{ .Values.podManagementPolicy }}
```

Comment:
Good thing, but I don't sure what is a default policy...

Reply:
> Good thing, but I don't sure what is a default policy...

Default of StatefulSet is to launch pods sequentially.
However chart installation would fail if I didn't add the value to `values.yaml`